### ISSUE TITLE
"cache" feels premature

### ISSUE THREAD

**Issue by annevk** [w3c/webappsec-clear-site-data#22]
We haven't quite explained how the HTTP cache sits in the browser. Some browsers have it double-keyed, some browsers key credentialed and non-credentialed differently because of CORS, ...

All browsers though share the cache to some extent across origins.

So offering a primitive to remove a chunk from that that potentially affects resources across origins strikes me as a little weird. It might not even help with attacks if the poisoned entry is in another origin.

**Comment by mikewest** [comment]
FWIW, `cache` is one of the main use cases that Google properties are interested in, along with the service worker kill switch. I guess I agree with you that it's poorly specified, but I'd like to improve the spec rather than remove the feature.

**Comment by annevk** [comment]
How would this even work for Google properties given that you delete an origin but Google has many domains? Can those folks weigh in directly with their scenarios? That might help.

**Comment by mikewest** [comment]
> How would this even work for Google properties given that you delete an origin but Google has many domains?

I don't think folks are interested in using this for `www.google.com`, nor are they interested in using the "`cookies`" bits. Those are blunt instruments, and wouldn't work well on a shared origin.

The properties that have expressed interest have their own subdomains (`photos.google.com`, `plus.google.com`, etc), and wish to clear cached user data upon logout. They'll fetch a resource from `lh3.googleusercontent.com` that responds with `Clear-Site-Data: cache`, we'll wipe stored images locally, and people will be happier.

> Can those folks weigh in directly with their scenarios? That might help.

I'll ask!

**Comment by mikewest** [comment]
(GitHub is already serving the header on logout, FWIW: https://twitter.com/mikewest/status/817391771508502529. Seems like a reasonably common use-case.)

**Comment by annevk** [comment]
So basically you ping all your CDNs to get their content uncached through a header and then uncache yourself through script? I guess it works if all that content is only used by you, but as soon as you're talking shared JavaScript libraries it gets icky.

**Comment by mikewest** [comment]
> So basically you ping all your CDNs to get their content uncached through a header and then uncache yourself through script?

Yes, except for the latter, where you uncache yourself via a header.

> I guess it works if all that content is only used by you, but as soon as you're talking shared JavaScript libraries it gets icky.

Yup. But remember that the target server is responsible for the cache-clearing response. I can't cause `cdn.yourserver.com` to remove all your cached JavaScript libraries; it has to make that decision in response to a request.

**Comment by annevk** [comment]
I guess overall it seems reasonable. Somewhat curious what @mnot would say about this and @mcmanus.

**Comment by mikewest** [comment]
In https://github.com/w3ctag/spec-reviews/issues/62, @mnot was mostly concerned about the cookie behavior. It was a year ago, but I think that was representative of most of the conversations at that point. Cache wasn't controversial, cookies (and `includeSubdomains`, which we removed) were. I think we've landed on something reasonable (if draconian) at this point, but I'm totally open to going around on it again.

**Comment by annevk** [comment]
Okay, I guess I'm fine with closing this then.

(I was wrong above by the way. Fetch also exposes parts of the HTTP cache through an API, but in a very different way.)

**Comment by mikewest** [comment]
I'll leave it open to give Mark and Patrick a chance to explain to me that I'm wrong. :)

**Comment by mnot** [comment]
I think clearing cache for a single origin (regardless of double keying, etc.) is reasonable -- assuming that's still what this does.

**Comment by mikewest** [comment]
That's still what this does. Thanks!

**Comment by slightlyoff** [comment]
A couple of thoughts from today's F2F in Tokyo (prepared with @travisleithead & @plinss):

 - Link to explainer and/or use-cases doc?
 - The names are a bit confusing at first glance: `cache` doesn't clear the SW Cache API. Maybe a different name for that one to indicate that this clears caches that aren't programmable?
 - Why is `storage` different to `cookies`? Is storage a superset? Conceptually cookies are a storage mechanism, so should other storage mechanisms be called out and be able to be reset granularly?
 - Is there a `*` value? Seems like that would be the default in an incident-response scenario. It that doesn't make sense, would love to see a note a to why
 - Are there are interactions with the credential manager API?
 
Thanks!

**Comment by annevk** [comment]
It seems like those be multiple new issues @slightlyoff. None of it is really related to OP.

**Comment by mikewest** [comment]
Yup. Not sure why you picked this bug to comment on instead of the review bug. :)

I'll answer inline, and we can figure out which things we should file new bugs for:

> Link to explainer and/or use-cases doc?

https://w3c.github.io/webappsec-clear-site-data/#intro is the best I've got, and it goes into a good amount of detail on both the overview and goals. It's not an explainer, but I'm hoping for some sort of grandfathered-in exception, since this has been kicking around in one form or another for years. :)

I'm happy to write something if there's still value, but it feels a bit like tacking on a design doc after building a project.

> The names are a bit confusing at first glance: cache doesn't clear the SW Cache API. Maybe a different name for that one to indicate that this clears caches that aren't programmable?

Sure. That is the distinction we're making, and if you have a naming suggestion, I'm happy to accept it.

> Why is storage different to cookies? Is storage a superset? Conceptually cookies are a storage mechanism, so should other storage mechanisms be called out and be able to be reset granularly?

Cookies don't respect the same-origin policy, so we have to clear them at the eTLD+1 level in order to leave a site in a sane state. I agree that the name `storage` could be improved. Help? :)

> Is there a * value? Seems like that would be the default in an incident-response scenario. It that doesn't make sense, would love to see a note a to why

We had `*` and removed it. I don't recall why, but adding it back is pretty trivial.

> Are there are interactions with the credential manager API?

No. We don't intend to give sites the ability to clear passwords, as those are in some way "user" data as opposed to "site" data. There's a good argument for including the `requires user mediation` bit in `cookies`, however. I'll make sure that's added.

**Comment by triblondon** [comment]
I have moved the TAG feedback into separate issues, following discussion at our recent F2F in London.  On the specific point of this issue, I think it would be helpful to explain more clearly what is covered by `cache` vs `storage`.  Reading the spec it seems very easy to place an individual data store in the wrong bucket.
